tico/passes/remove_redundant_reshape.py

```diff
@@ -427,3 +428,113 @@ def call(self, exported_program: ExportedProgram) -> PassResult:
         graph_module.recompile()
 
         return PassResult(modified)
+
+
+@trace_graph_diff_on_pass
+class RemoveRedundantReshapePattern6(PassBase):
+    def __init__(self):
+        super().__init__()
+        self.SHAPE_PRESERVING_OPS = [
+            *ops.aten.mul_scalar,
+            *ops.aten.mul_tensor,
+            *ops.aten.add,
+            # TODO add more ops.
+        ]
+
+    def call(self, ep: ExportedProgram) -> PassResult:
+        """
+        Elide a redundant leading-dimension unsqueeze that is immediately
+        cancelled later.
+
+        [BEFORE]
+          x: (20x1x512)
+          R1 = aten.reshape(x, shape=(1, 20, 8, 64))  # adds a leading 1
+          P = aten.permute (R1, dims=(0, 2, 1, 3))    # keeps that 1 in front
+            ... arbitrary shape-preserving ops ...
+          R2 = aten.reshape(*, shape=(8, 20, 64))     # drops the leading 1
+
+        [AFTER]
+          R1 = aten.reshape(x, shape=(20, 8, 64))
+          P = aten.permute (R1, dims=(1, 0, 2))
+            ... arbitrary shape-preserving ops ...
+          R2 = aten.reshape(*, shape=(8, 20, 64))     # same as before
+
+        Effect:
+            * The first reshape loses one unit-dim.
+            * The permute dims are shifted down by 1 and the leading 0 is removed.
+        """
+        logger = logging.getLogger(__name__)
+
+        gm = ep.graph_module
+        graph = gm.graph
+        modified = False
+        for reshape_front in graph.nodes:
+            if not is_target_node(reshape_front, ops.aten.reshape):
+                continue
+            if len(reshape_front.users) != 1:
+                continue
+            # The reshape must actually add a leading 1.
+            reshape_front_shape = extract_shape(reshape_front)
+            if len(reshape_front_shape) != 0 and reshape_front_shape[0] != 1:
+                continue
+
+            permute = next(iter(reshape_front.users))
+            if not is_target_node(permute, ops.aten.permute):
+                continue
+
+            # The permute must keep that leading dim in front.
+            permute_shape = extract_shape(permute)
+            if permute_shape[0] != 1:
```

Comment:
May I ask a question? 
Could you explain Why we check `permute_shape[0]` == 1?

Reply:
> The permute must keep that leading dim in front.

This is related with what this pass does.